src/memray/_memray/record_reader.cpp

```diff
@@ -243,22 +254,42 @@ RecordReader::parseNativeAllocationRecord()
 }
 
 bool
-RecordReader::parseSegmentHeader()
-{
-    std::string filename;
-    uintptr_t addr;
-    size_t num_segments;
-    if (!d_input->getline(filename, '\0')
-        || !d_input->read(reinterpret_cast<char*>(&num_segments), sizeof(num_segments))
-        || !d_input->read(reinterpret_cast<char*>(&addr), sizeof(addr)))
-    {
-        return false;
-    }
+RecordReader::parseMemoryMapStart()
+{
+    // Currently nothing to do (this record type has no body)
+    return true;
+}
+
+bool
+RecordReader::processMemoryMapStart()
+{
+    std::lock_guard<std::mutex> lock(d_mutex);
+    d_symbol_resolver.clearSegments();
+    return true;
+}
+
+bool
+RecordReader::parseSegmentHeader(std::string* filename, size_t* num_segments, uintptr_t* addr)
+{
+    return d_input->getline(*filename, '\0')
+           && d_input->read(reinterpret_cast<char*>(num_segments), sizeof(*num_segments))
+           && d_input->read(reinterpret_cast<char*>(addr), sizeof(*addr));
+}
 
+bool
+RecordReader::processSegmentHeader(const std::string& filename, size_t num_segments, uintptr_t addr)
+{
     std::vector<Segment> segments(num_segments);
     for (size_t i = 0; i < num_segments; i++) {
+        RecordType record_type;
+        if (!d_input->read(reinterpret_cast<char*>(&record_type), sizeof(record_type))
```

Comment:
Technically this is *parsing* instead of *processing*, no? I don't think is a problem and not sure if is worth moving this into another function, but is a bit surprising given the pattern you are entablishing.

To be honest, I would do nothing here but wanted to point out in case you want to discuss :)

Reply:
Yeah, it's a bit weird, but I couldn't come up with anything nicer. The `dumpAllRecords` function wants to just parse segment records as it encounters them, and the `nextRecord()` function wants to greedily slurp up every `SEGMENT` record it encounters immediately after a `SEGMENT_HEADER`. The structure I'm moving it to in this PR is the simplest one I could come up with that retains the existing behavior for `nextRecord` (enter a mode where only `SEGMENT` records are accepted) and for `dumpAllRecords` (don't do that). If you see a nicer option, I'm all ears.

Reply:
Right. The only think I can think of is to factor this part into `parseSegments()` where the caller passes the fector and passing the vector in to this function that calls `d_symbol_resolver.addSegments`

Reply:
I'm just gonna leave this alone. I agree that it could be nicer, but I think it's "ok" as it is, and anything else seems like it would be more complex. We can always clean this up further in another PR.